libraries/chain/fork_database.cpp

```diff
@@ -23,27 +32,31 @@ namespace eosio { namespace chain {
       indexed_by<
          hashed_unique< tag<by_block_id>, member<block_header_state, block_id_type, &block_header_state::id>, std::hash<block_id_type>>,
          ordered_non_unique< tag<by_prev>, const_mem_fun<block_header_state, const block_id_type&, &block_header_state::prev> >,
-         ordered_non_unique< tag<by_block_num>,
+         ordered_unique< tag<by_lib_block_num>,
             composite_key< block_state,
-               member<block_header_state,uint32_t,&block_header_state::block_num>,
-               member<block_state,bool,&block_state::in_current_chain>
-            >,
-            composite_key_compare< std::less<uint32_t>, std::greater<bool> >
-         >,
-         ordered_non_unique< tag<by_lib_block_num>,
-            composite_key< block_header_state,
-                member<block_header_state,uint32_t,&block_header_state::dpos_irreversible_blocknum>,
-                member<block_header_state,uint32_t,&block_header_state::bft_irreversible_blocknum>,
-                member<block_header_state,uint32_t,&block_header_state::block_num>
+               member<block_state,                        bool,          &block_state::validated>,
+               member<detail::block_header_state_common, uint32_t,      &detail::block_header_state_common::dpos_irreversible_blocknum>,
+               member<detail::block_header_state_common, uint32_t,      &detail::block_header_state_common::block_num>,
+               member<block_header_state,                block_id_type, &block_header_state::id>
             >,
-            composite_key_compare< std::greater<uint32_t>, std::greater<uint32_t>, std::greater<uint32_t> >
+            composite_key_compare<
+               std::greater<bool>,
+               std::greater<uint32_t>,
+               std::greater<uint32_t>,
+               std::less<block_id_type>
+            >
```

Comment:
Since replay is needed anyway, this is a good time to use the faster less for sha256:
```
   struct sha256_less {
      bool operator()( const sha256& lhs, const sha256& rhs ) const {
       return
             std::tie(lhs._hash[0], lhs._hash[1], lhs._hash[2], lhs._hash[3]) <
             std::tie(rhs._hash[0], rhs._hash[1], rhs._hash[2], rhs._hash[3]);
      }
   };
```